### DISCUSSION TITLE
Calculating percentages (survey-html-form)

### DISCUSSION THREAD

**Discussion by CGaele** [jspsych/jsPsych#1224] [Q&A]
Hello,

As part of a questionnaire, we ask participants to provide three different percentages for every item in a list of 14 questions. The questions are something like: what language(s) do you speak at home? Then they indicate a percentage for Chinese, English and 'other language(s)', which must add up to 100%. The idea is that (perc-chi-use1+perc-eng-use1+perc-other-use1)=100% (perc-chi-use2+perc-eng-use2+perc-other-use2)=100%, etc etc. If more or less than 100%, we don't want them to move on to the next page (similar to 'required'). I tried coding in different ways using conditionals, but they don't seem to be working.

Many thanks in advance,
Claudia

        var section6_languse = {
            type: 'survey-html-form',
            preamble: '<br><strong>请您提供您在不同年纪与不同对象，使用中文或中文以外的比例。使用百分来估计语言使用的时间。'+
                      '例如：如果您只使用中文且不曾使用过英语或其他语言，请在中文栏填写100，英文栏填写0，其他栏为空白。</strong>',
            html: '<div class="lb-question-container"><ol type="1">'+
                    '<li>您在6岁以前，在家里使用:</p><p>中文 <input type="number" name="q1" id="perc-chi-use1" class="perc" min="0" max="100" maxlength="3" required>%</p><p>英语<input type="number" name="q1" id="perc-eng-use1" class="perc" min="0" max="100" maxlength="3" required>%</p><p>其他（请注明）： <input type="text" name="other1" id="other-lang1" autocomplete="off"></label>使用： <input type="number" name="q1" id="perc-other-use1" class="perc" min="0" max="100" maxlength="3">%<br><br></li>'+
                    '<li>您在6 岁和12岁之间，在家里使用:</p><p>中文 <input type="number" name="q2" id="perc-chi-use2" class="perc" min="0" max="100" maxlength="3" required>%</p><p>英语 <input type="number" name="q2" id="perc-eng-use2" class="perc" min="0" max="100" maxlength="3" required>%</p><p>其他（请注明）： <input type="text" name="other2" id="other-lang-use2" autocomplete="off"></label>使用： <input type="number" name="q2" id="perc-other-use2" class="perc" min="0" max="100" maxlength="3">%<br><br></li>'+
                    '<li>您在13 岁和15岁之间，在家里使用:</p><p>中文 <input type="number" name="q3" id="perc-chi-use3" class="perc" min="0" max="100" maxlength="3" required>%</p><p>英语 <input type="number" name="q3" id="perc-eng-use3" class="perc" min="0" max="100" maxlength="3" required>%</p><p>其他（请注明）： <input type="text" name="other3" id="other-lang-use3" autocomplete="off"></label>使用： <input type="number" name="q3" id="perc-other-use3" class="perc" min="0" max="100" maxlength="3">%<br><br></li>'+
                    '<li>您在过去5年内，在家里使用:</p><p>中文 <input type="number" name="q4" id="perc-chi-use4" class="perc" min="0" max="100" maxlength="3" required>%</p><p>英语 <input type="number" name="q4" id="perc-eng-use4" class="perc" min="0" max="100" maxlength="3" required>%</p><p>其他（请注明）： <input type="text" name="other4" id="other-lang-use4" autocomplete="off"></label>使用： <input type="number" name="q4" id="perc-other-use4" class="perc" min="0" max="100" maxlength="3">%<br><br></li>'+
                    '<li>您在6岁以前，在学校使用:</p><p>中文 <input type="number" name="q5" id="perc-chi-use5" class="perc" min="0" max="100" maxlength="3" required>%</p><p>英语 <input type="number" name="q5" id="perc-eng-use5" class="perc" min="0" max="100" maxlength="3" required>%</p><p>其他（请注明）： <input type="text" name="other5" id="other-lang-use5" autocomplete="off"></label>使用：<input type="number" name="q5" id="perc-other-use5" class="perc" min="0" max="100" maxlength="3">%<br><br></li>'+
                    '<li>您在6 岁和12岁之间，在学校使用:</p><p>中文 <input type="number" name="q6" id="perc-chi-use6" class="perc" min="0" max="100" maxlength="3" required>%</p><p>英语 <input type="number" name="q6" id="perc-eng-use6" class="perc" min="0" max="100" maxlength="3" required>%</p><p>其他（请注明）： <input type="text" name="other6" id="other-lang-use6" autocomplete="off"></label>使用：<input type="number" name="q6" id="perc-other-use6" class="perc" min="0" max="100" maxlength="3">%<br><br></li>'+
                    '<li>您在13 岁和15岁之间，在家里或工作场合使用:</p><p>中文 <input type="number" name="q7" id="perc-chi-use7" class="perc" min="0" max="100" maxlength="3" required>%</p><p>英语 <input type="number" name="q7" id="perc-eng-use7" class="perc" min="0" max="100" maxlength="3" required>%</p><p>其他（请注明）： <input type="text" name="other7" id="other-lang-use7" autocomplete="off"></label>使用：<input type="number" name="q7" id="perc-other-use7" class="perc" min="0" max="100" maxlength="3">%<br><br></li>'+
                    '<li>您在过去5年内，在家里或工作场合使用:</p><p>中文 <input type="number" name="q8" id="perc-chi-use8" class="perc" min="0" max="100" maxlength="3" required>%</p><p>英语 <input type="number" name="q8" id="perc-eng-use8" class="perc" min="0" max="100" maxlength="3" required>%</p><p>其他（请注明）： <input type="text" name="other8" id="other-lang-use8" autocomplete="off"></label>使用：<input type="number" name="q8" id="perc-other-use8" class="perc" min="0" max="100" maxlength="3">%<br><br></li>'+
                    '<li>您在过去5年内，当您在与您的家人一起时:</p><p>中文 <input type="number" name="q9" id="perc-chi-use9" class="perc" min="0" max="100" maxlength="3" required>%</p><p>英语 <input type="number" name="q9" id="perc-eng-use9" class="perc" min="0" max="100" maxlength="3" required>%</p><p>其他（请注明）： <input type="text" name="other9" id="other-lang-use9" autocomplete="off"></label>使用：<input type="number" name="q9" id="perc-other-use9" class="perc" min="0" max="100" maxlength="3">%<br><br></li>'+
                    '<li>您在过去5年内，当您和您的朋友在一起时:</p><p>中文 <input type="number" name="q10" id="perc-chi-use10" class="perc" min="0" max="100" maxlength="3" required>%</p><p>英语 <input type="number" name="q10" id="perc-eng-use10" class="perc" min="0" max="100" maxlength="3" required>%</p><p>其他（请注明）： <input type="text" name="other10" id="other-lang-use10" autocomplete="off"></label>使用：<input type="number" name="q10" id="perc-other-use10" class="perc" min="0" max="100" maxlength="3">%<br><br></li>'+
                    '<li>您在过去5年内，当您在思考时:</p><p>中文 <input type="number" name="q11" id="perc-chi-use11" class="perc" min="0" max="100" maxlength="3" required>%</p><p>英语 <input type="number" name="q11" id="perc-eng-use11" class="perc" min="0" max="100" maxlength="3" required>%</p><p>其他（请注明）： <input type="text" name="other11" id="other-lang-use11" autocomplete="off"></label>使用：<input type="number" name="q11" id="perc-other-use11" class="perc" min="0" max="100" maxlength="3">%<br><br></li>'+
                    '<li>您在过去5年内，当您阅读文字/书籍/报纸:</p><p>中文 <input type="number" name="q12" id="perc-chi-use12" class="perc" min="0" max="100" maxlength="3" required>%</p><p>英语 <input type="number" name="q12" id="perc-eng-use12" class="perc" min="0" max="100" maxlength="3" required>%</p><p>其他（请注明）： <input type="text" name="other12" id="other-lang-use12" autocomplete="off"></label>使用：<input type="number" name="q12" id="perc-other-use12" class="perc" min="0" max="100" maxlength="3">%<br><br></li>'+
                    '<li>您在过去5年内，当您看电视和/或在收听广播时:</p><p>中文 <input type="number" name="q13" id="perc-chi-use13" class="perc" min="0" max="100" maxlength="3" required>%</p><p>英语 <input type="number" name="q13" id="perc-eng-use13" class="perc" min="0" max="100" maxlength="3" required>%</p><p>其他（请注明）： <input type="text" name="other13" id="other-lang-use13" autocomplete="off"></label>使用：<input type="number" name="q13" id="perc-other-use13" class="perc" min="0" max="100" maxlength="3">%<br><br></li>'+
                    '<li>您在过去5年内，与您的伴侣在一起时. 假如您没有伴侣，请填NA: </p><p>中文 <input type="number" name="q14" id="perc-chi-use14" class="perc" min="0" max="100" maxlength="3" required>%</p><p>英语 <input type="number" name="q14" id="perc-eng-use14" class="perc" min="0" max="100" maxlength="3" required>%</p><p>其他（请注明）： <input type="text" name="other14" id="other-lang-use14" autocomplete="off"></label>使用：<input type="number" name="q14" id="perc-other-use14" class="perc" min="0" max="100" maxlength="3">%</p></li>'+
                    '</ol></div>',
        };
        timeline.push(section6_languse)

**Comment by jodeleeuw** [comment]
Hi Claudia,

Unfortunately we don't yet have custom validation functions implemented, but it's in the works. 

But you can use the `on_load` event to setup custom event handlers to do this kind of validation.

I'm going to provide an example that's a simplification of your trials, but I hope illustrates how you could do this:

```js
var trial = {
  type: 'survey-html-form',
  html: `<ul>
    <li>Option 1: <input value="0" type="number" name="q1" id="u1" min="0" max="100" maxlength=3 required></input></li>
    <li>Option 2: <input value="0" type="number" name="q2" id="u2" min="0" max="100" maxlength=3 required></input></li>
    <li>Option 3: <input value="0" type="number" name="q3" id="u3" min="0" max="100" maxlength=3 required></input></li>
  </ul>
  <p id="status">All three options must add up to 100.</p>
  `,
  on_load: function(){
    document.querySelector('input[type="submit"]').disabled = true;
    var inputs = document.querySelectorAll('input[type="number"]');
    for(var i=0; i<inputs.length; i++){
      inputs[i].addEventListener('change', function(){
        if(checkSum()){
          document.querySelector('input[type="submit"]').disabled = false;
          document.querySelector('#status').style.visibility = 'hidden';
        } else {
          document.querySelector('input[type="submit"]').disabled = true;
          document.querySelector('#status').style.visibility = 'visible';
        }
      })
    }
    
    function checkSum(){
      var sum = (
        parseInt(document.querySelector('#u1').value) +
        parseInt(document.querySelector('#u2').value) +
        parseInt(document.querySelector('#u3').value)
      )
      console.log(sum);
      return sum == 100
    }
    
  }
}
```

You can try it out here: https://xdum7vjqza.cognition.run/
